myhpom/tasks.py

```diff
@@ -0,0 +1,119 @@
+from django.conf import settings
+from django.template.loader import get_template
+from django.template import Context
+from celery import shared_task
+from celery.task import Task
+from celery.signals import task_failure
+from myhpom.models import AdvanceDirective, DocumentUrl, CloudFactoryRun, CloudFactoryUnit
+
+
+@shared_task
+class CloudFactorySubmitAdvanceDirectiveRun(Task):
+    def run(self, ad_id, line_id, document_host, callback_url):
+        """
+        ## Arguments
+        * ad_id = the database id for the AdvanceDirective
+        * line_id = the id of the CloudFactory production line to submit this request to.
+        * document_host = the host (incl. protocol/scheme) to use for the DocumentURL
+        * callback_url = the URL to use with the callback request
+
+        ## Task Process
+        - create an expiring DocumentUrl for the AdvanceDirective
+        - submit the DocumentUrl to the CloudFactory (test) endpoint
+        - store the status of the submitted document
+        - if the AdvanceDirective no longer exists, abort the task without an Exception.
+            - the AdvanceDirective might no longer exist by the time celery picks it up,
+            - or the document it points to might have been removed.
+        - if an error occurs in this process, send support email.
+        """
+        cf_run = CloudFactoryRun.objects.create(line_id=line_id, callback_url=callback_url or "")
+
+        # If the AdvanceDirective has been deleted, abort the task with a message
+        # -- no need to admin support email, this is an expected case.
+        try:
+            ad = AdvanceDirective.objects.get(id=ad_id)
+        except:
+            cf_run.status = 'ABORTED'
+            cf_run.message = 'AdvanceDirective(id=%r) no longer exists.' % ad_id
+            cf_run.save()
+            return cf_run.data
+
+        # Any exception in the rest of the task should result in support email
+        document_url = DocumentUrl.objects.create(advancedirective=ad)
+        CloudFactoryUnit.objects.create(
+            run=cf_run, input=self.create_unit_input(document_url, document_host)
+        )
+        response = self.post_run(cf_run.post_data)
+        response_data = response.json()
+
+        # if the run could not be created (status_code != 201), send support email
+        # -- we need to understand why the the run could not be created at CloudFactory.
+        if response.status_code != 201:
+            cf_run.__dict__.update(
+                status=str(response.status_code), message=response_data.get('message')
+            )
+            cf_run.save()
+            raise ValueError(
+                "%s\n\n== RUN POST DATA ==\n%s\n\n== FULL RUN DATA ==\n%s"
+                % (
+                    response_data['message'],
+                    json.dumps(cf_run.post_data, indent=2),
+                    json.dumps(cf_run.data, indent=2),
+                )
+            )
+
+        cf_run.__dict__.update(
```

Comment:
This `model.__dict__.update` pattern doesn't feel very pythonic, and I don't think it has any advantage over a an explicit `model.field = value` pattern.

That said this does remind me that tasks should be somewhat atomic, or as atomic as we can make them.

One common pattern to prevent any potential 'partial updates' from other SQL requests is to do:

```CloudFactoryRun.objects.filter(pk=cf_run.pk).update(...all the fileds)```

This has the advantage that one query is made to both select the latest values in the database (locked to prevent others from moding that record), and updating the records we care about.

I'm not sure what 'atomic' means for this task - probably NOT wrapping the whole thing in a sql atomic block...yeah, I think it might just be the filter/update pattern I suggest above. But any thoughts on that welcome (the celery docs go into length about chaining tasks to deal so that tasks do one thing its atomic/idempotent-as-possible, etc...I don't know that we need to go that far, but that too might be a good excersize...)

Reply:
Django not providing variable object[key] field access is pretty unpythonic to me, which is why I keep using the __dict__ object. But I like your solution, to do it in an atomic .update().